### DISCUSSION TITLE
How to create a workflow instance without immediately executing it and execut it later?

### DISCUSSION THREAD

**Discussion by coolshun** [elsa-workflows/elsa-core#5133] [Q&A]
How to create a workflow instance without immediately executing it and execute it later,with elsa3.0?I reqeust the api "/workflow-definitions/{workflow_definition_id}/dispatch" to create a workflow instance.And then I want to request the api "/workflow-instances/{workflow_instance_id}/execute" to execute the instance,but I can't find the api in the directory "elsa-core\src\modules\Elsa.Workflows.Api\Endpoints\WorkflowInstances\".I using Elsa 3.0.

**Comment by glime-ai** [comment]
To create a workflow instance without immediately executing it in Elsa 3.0, you can use the `/workflow-definitions/{workflow_definition_id}/dispatch` endpoint. This will enqueue the workflow for execution and return the workflow instance ID, which you can use later to track or manage the workflow instance.

However, based on the context provided, it seems that you are looking for an endpoint to execute a workflow instance by its ID, such as `/workflow-instances/{workflow_instance_id}/execute`, but you cannot find it in the directory `elsa-core\src\modules\Elsa.Workflows.Api\Endpoints\WorkflowInstances\`.

As of the information provided, there is no direct mention of an endpoint to execute an already created workflow instance by its ID in the provided documentation. This might be an oversight in the documentation, a feature not yet implemented, or simply not exposed via the API in the version you are using.

To proceed, you could:

1. Check the official Elsa documentation or GitHub repository for any updates or additional information regarding the execution of workflow instances by ID.
2. Look into the possibility of using the `IWorkflowInstanceExecutor` service directly in your code to execute a workflow instance if you have access to the server-side code.
3. Reach out to the Elsa community or support channels for guidance on how to execute a workflow instance by ID using the REST API or other means.

If the feature is indeed missing from the API, you may need to implement custom logic to handle the execution of workflow instances or wait for a future release that includes this functionality.

**Comment by coolshun** [comment]
@noxe @petedavis @lahma @cando

**Comment by coolshun** [comment]
The api "/workflow-definitions/{workflow_definition_id}/dispatch" not just creating instance without executing it,still executing it .How to create a workflow instance without immediately executing it and execut it later? @noxe @petedavis @lahma @cando

**Comment by sfmskywalker** [comment]
Hi @coolshun, there's currently no API endpoint to create a workflow instance and another endpoint to start it. But, it should be relatively easy to add. If you want, you can open a new issue with this feature request. Please also include details about your use case scenario. which will be helpful.

**Reply by coolshun** [reply]
What we are doing is a disaster recovery exercise system, which is part of BCM. Design the exercise workflow definiton first, and then create a workflow instance. At this point, the status of instance  is "NotStarted". And  a dashboard will display information about the instance, such as flowchart, status, number of tasks, etc. Many people are watching it. Then start the workflow instance and display the execution process  in real-time on the dashboard.So I want to create a workflow instance without immediately executing it and execut it later.

**Comment by glime-ai** [comment]
To create a workflow instance without immediately executing it and to execute it later in Elsa 3.0, you can use the `/workflow-definitions/{workflow_definition_id}/dispatch` endpoint to create the workflow instance. This endpoint will enqueue the workflow for execution and return the workflow instance ID, which you can use to track the workflow instance.

However, based on the information provided, it seems that the `/workflow-instances/{workflow_instance_id}/execute` endpoint you are looking for does not exist in the directory `elsa-core\src\modules\Elsa.Workflows.Api\Endpoints\WorkflowInstances\`. This could mean that the functionality to execute a workflow instance manually by an API call might not be available in Elsa 3.0 as of the provided context.

If you need to execute the workflow instance later, you might need to look into the available services that Elsa provides for managing workflow instances programmatically. For example, you could use the `IWorkflowInstanceExecutor` service if it's available in Elsa 3.0 to execute a workflow instance manually from your code.

Since the direct API endpoint is not available, you may need to check the official Elsa documentation or reach out to the Elsa community or support for guidance on how to achieve this with the latest version of Elsa.

If you have access to the source code or the ability to extend the API, you could potentially implement a custom endpoint that allows you to execute a workflow instance by its ID.
